js/msg.js

```diff
@@ -130,27 +130,30 @@ window.INJECTED !== 1 && (() => {
     },
   };
 
-  window.API = new Proxy({}, {
-    get(target, name) {
-      // using a named function for convenience when debugging
-      return async function invokeAPI(...args) {
-        if (!bg && chrome.tabs) {
-          bg = await browser.runtime.getBackgroundPage().catch(() => {});
-        }
-        const message = {method: 'invokeAPI', name, args};
-        // content scripts and probably private tabs
-        if (!bg) {
-          return msg.send(message);
-        }
-        // in FF, the object would become a dead object when the window
-        // is closed, so we have to clone the object into background.
-        const res = bg.msg._execute(TARGETS.extension, bg.deepCopy(message), {
-          frameId: 0, // false in case of our Options frame but we really want to fetch styles early
-          tab: NEEDS_TAB_IN_SENDER.includes(name) && await getOwnTab(),
-          url: location.href,
-        });
-        return deepCopy(await res);
-      };
+  const apiHandler = !isBg && {
+    get({PATH}, name) {
+      const fn = () => {};
+      fn.PATH = [...PATH, name];
+      return new Proxy(fn, apiHandler);
     },
-  });
+    async apply({PATH: path}, thisObj, args) {
+      if (!bg && chrome.tabs) {
+        bg = await browser.runtime.getBackgroundPage().catch(() => {});
+      }
+      const message = {method: 'invokeAPI', path, args};
+      // content scripts and probably private tabs
+      if (!bg) {
+        return msg.send(message);
+      }
+      // in FF, the object would become a dead object when the window
+      // is closed, so we have to clone the object into background.
+      const res = bg.msg._execute(TARGETS.extension, bg.deepCopy(message), {
+        frameId: 0, // false in case of our Options frame but we really want to fetch styles early
+        tab: NEEDS_TAB_IN_SENDER.includes(path.join('.')) && await getOwnTab(),
+        url: location.href,
+      });
+      return deepCopy(await res);
+    },
+  };
+  window.API = isBg ? {} : new Proxy({PATH: []}, apiHandler);
```

Comment:
I think we can split out `api` stuff to a new file like `js/api.js`:
```js
/* global msg */

const API = () => {
  return msg.isBg ? createServerAPI() : createClientAPI(); // if this confuses the editor, it might be better to split them into api-client.js and api-server.js
  
  function createServerAPI() {
     ... 
     return {
      register: (name, methods) => {...}
    };
  }
  
  function createClientAPI() {
     // ...
     return new Proxy({}, apiHandler);
  }
}();
```

For clients, just call API as usual (with category); for services (e.g. styleManager or other singletons):
```js
/* global API ... */
const styleManager = (() => {
  ...
})();
API.register('styles', styleManger);
```

Reply:
I don't like `API.register` because 1) it looks like an API method by itself, 2) it makes everything indirect again so I can't `jump to definition` and `peek source` in IDE, 3) there's no need to register anything as it's just a direct assignment and we're not trying to reimplement module system like `require()`, 4) I don't like adding more files to content scripts because every additional file needlessly increases page load time - first we should bundle our content scripts into one at least in production zip.

Reply:
Oh I think I got your point. You are trying to use `jump to definition` to navigate from e.g. https://github.com/openstyles/stylus/blob/cb64a6bac96dc6a61b9860c1ee3af09bc0fdb803/content/apply.js#L60 to https://github.com/openstyles/stylus/blob/cb64a6bac96dc6a61b9860c1ee3af09bc0fdb803/background/style-via-api.js#L26 ?

Merging all function into a single namespace does make sense now. I still don't think it is a good idea though.

Reply:
I want to switch to a structured API namespace. For example like `chrome` or `browser` namespace. Our previous API was a flat list with a lot of indirection, duplication, aliasing.

>You are trying to use jump to definition to navigate 

Yes. That's one of the main conveniences an IDE provides.

Reply:
FYI, when `Worker` is introduced, some Google guys also tried to solve these types of problems (share the same object shape in two different scopes/navigation) in a similar way. They do something like this:

*server.js*
```js
class MyService {
  myMethod() { ... }
}

exposeAPI(MyService);
```
*client.js*
```js
/* magic comment that makes the IDE think this has the same shape as `MyService` in server.js */
const MyService = useAPI('MyService');

const service = new MyService();
await service.myMethod();
```
https://github.com/GoogleChromeLabs/comlink

Reply:
This is one of the possible methods, yes. It works because they don't explicitly duplicate method names in two places like we do currently to enable automatic loadScript.